packages/x-data-grid/src/hooks/features/rowSelection/utils.ts

```diff
@@ -13,3 +118,137 @@ export function isMultipleRowSelectionEnabled(
+  apiRef: React.MutableRefObject<GridPrivateApiCommunity>,
+  tree: GridRowTreeConfig,
+  selectedRow: GridRowId,
+  autoSelectDescendants: boolean,
+  autoSelectParents: boolean,
+) => {
+  const filteredRows = gridFilteredRowsLookupSelector(apiRef);
+  const selectedIdsLookup = selectedIdsLookupSelector(apiRef);
+  const rowsToSelect: Set<GridRowId> = new Set([]);
+
+  if (!autoSelectDescendants && !autoSelectParents) {
+    return rowsToSelect;
+  }
+
+  if (autoSelectDescendants) {
+    const rowNode = tree[selectedRow];
+
+    if (rowNode?.type === 'group') {
+      const rowGroupDescendantsSelector = getGridRowGroupSelectableDescendantsSelector(
+        apiRef,
+        selectedRow,
+      );
+      const descendants = rowGroupDescendantsSelector(apiRef);
+      descendants.forEach(rowsToSelect.add, rowsToSelect);
+    }
+  }
+
+  if (autoSelectParents) {
+    const checkAllDescendantsSelected = (rowId: GridRowId): boolean => {
+      if (selectedIdsLookup[rowId] !== rowId && !rowsToSelect.has(rowId)) {
+        return false;
+      }
+      const node = tree[rowId];
+      if (node?.type !== 'group') {
+        return true;
+      }
+      return node.children.every((child) => checkAllDescendantsSelected(child));
+    };
+
+    const traverseParents = (rowId: GridRowId) => {
+      const siblings: GridRowId[] = getFilteredRowNodeSiblings(tree, filteredRows, rowId);
+      if (
+        siblings.length === 0 ||
+        siblings.every((sibling) => checkAllDescendantsSelected(sibling))
+      ) {
+        const rowNode = tree[rowId] as GridGroupNode;
+        const parent = rowNode.parent;
+        if (isValidRowId(parent) && parent !== GRID_ROOT_GROUP_ID && apiRef.current.isRowSelectable(parent)) {
+          rowsToSelect.add(parent);
+          traverseParents(parent);
+        }
+      }
+    };
+    traverseParents(selectedRow);
+  }
+
+  return Array.from(rowsToSelect);
+};
+
+export const findRowsToDeselect = (
+  apiRef: React.MutableRefObject<GridPrivateApiCommunity>,
+  tree: GridRowTreeConfig,
+  deselectedRow: GridRowId,
+  autoSelectDescendants: boolean,
+  autoSelectParents: boolean,
+) => {
+  const rowsToDeselect: GridRowId[] = [];
+  const selectedIdsLookup = selectedIdsLookupSelector(apiRef);
+
+  if (!autoSelectParents && !autoSelectDescendants) {
+    return rowsToDeselect;
+  }
+
+  if (autoSelectParents) {
+    const allParents = getRowNodeParents(tree, deselectedRow);
+    allParents.forEach((parent) => {
+      const isSelected = selectedIdsLookup[parent] === parent;
+      if (isSelected) {
+        rowsToDeselect.push(parent);
+      }
+    });
+  }
+
+  if (autoSelectDescendants) {
+    const rowNode = tree[deselectedRow];
+    if (rowNode?.type === 'group') {
+      const rowGroupDescendantsSelector = getGridRowGroupSelectableDescendantsSelector(
+        apiRef,
+        deselectedRow,
+      );
+      const descendants = rowGroupDescendantsSelector(apiRef);
+      return rowsToDeselect.concat(descendants);
```

Comment:
`forEach`+`push` is usually faster than `concat`: https://jsperf.app/jozuna/1/preview